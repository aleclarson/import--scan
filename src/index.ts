export function importScan(code: string) {
  let imports: string[] = []
  let lexer = /(\/\/.*$)|(\/\*)|(?:^[ \t]*import(?: [^\n]+? from)?[ ]+["'`]([^"'`]+)["'`])|(?:(?<=[^\w.$])(?:require|import)\(\s*["'`]([^"'`]+)["'`]\s*\))/gm
  while (true) {
    let match = lexer.exec(code)
    if (!match) return imports

    // Line comment
    if (match[1]) continue

    // Block comment
    if (match[2]) {
      let start = match.index + 2
      while (true) {
        let end = code.indexOf('*/', start)
        if (end < 0) return imports
        if (isEscaped(code, end)) {
          start = end + 1
          continue
        }
        lexer.lastIndex = end + 2
        break
      }
    }

    // Imported path
    else {
      let path = match[3] || match[4]
      if (imports.indexOf(path) < 0) imports.push(path)
    }
  }
}

// Returns true when the given string ends with an unescaped escape.
function isEscaped(str: string, fromIndex: number) {
  let ESCAPE = '\\'.charCodeAt(0),
    i = fromIndex,
    n = 0
  while (i && str.charCodeAt(--i) === ESCAPE) n++
  return n % 2 == 1
}
