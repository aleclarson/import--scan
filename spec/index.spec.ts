import { getImports } from '../src'

test('import statements', () => {
  expect(
    getImports(`
      import xyz from '1';
      import * as name from \`2\`;
      import { foo } from "3";
      import { foo as bar } from "4";
      import { foo, bar } from "5";
      import { foo, bar as bar2 } from "6";
      import xyz, { foo } from "7";
      import xyz, * as foo from "8";
      import '9';
    `)
  ).toMatchSnapshot()
})

test('import calls', () => {
  expect(
    getImports(`
      let loadFoo = () => import("foo")
      let barPromise = import('bar');
      import(\`event-stream\`)
    `)
  ).toMatchSnapshot()
})

test('require calls', () => {
  expect(
    getImports(`
      let bar = require('bar');
      let foo = require("foo")
      require(\`event-stream\`);

      // False positives
      foo.require('xyz')
      $require('xyz')
      _require('xyz')
    `)
  ).toMatchSnapshot()
})

test('mixed import and require', () => {
  expect(
    getImports(`
      import '1';
      require('2');
      import('3');
    `)
  ).toMatchSnapshot()
})

test('"from" identifier in import statement', () => {
  expect(getImports(`import { from } from 'from'`)).toMatchSnapshot()
})

test('shared path between imports', () => {
  expect(
    getImports(`
      import { foo1 } from 'foo'
      import { bar } from 'bar'
      import { foo2 } from 'foo'
    `)
  ).toMatchSnapshot()
})

test('line comments', () => {
  expect(
    getImports(`
      // require('foo')
      // import bar from 'bar'
    `)
  ).toMatchSnapshot()
})

test('block comments', () => {
  expect(
    getImports(`
      /* require('foo') */
      /* import xyz from 'xyz' */
      /*
      require('bar');
      *//*
      import xyz from 'xyz'; */
    `)
  ).toMatchSnapshot()
})

test('escaped block comment', () => {
  expect(
    getImports(`
      /* \\*/ require('1') */
      /* \\\\*/ require('2')
      /* \\\\\\*/ require('3') */
    `)
  ).toMatchSnapshot()
})

test('unclosed block comment', () => {
  expect(getImports('/* import foo from "foo"')).toMatchSnapshot()
})
