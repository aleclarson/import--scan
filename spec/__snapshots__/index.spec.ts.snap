// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"from" identifier in import statement 1`] = `
Array [
  "from",
]
`;

exports[`block comments 1`] = `Array []`;

exports[`escaped block comment 1`] = `
Array [
  "2",
]
`;

exports[`import calls 1`] = `
Array [
  "foo",
  "bar",
  "event-stream",
]
`;

exports[`import statements 1`] = `
Array [
  "1",
  "2",
  "3",
  "4",
  "5",
  "6",
  "7",
  "8",
  "9",
]
`;

exports[`line comments 1`] = `Array []`;

exports[`mixed import and require 1`] = `
Array [
  "1",
  "2",
  "3",
]
`;

exports[`require calls 1`] = `
Array [
  "bar",
  "foo",
  "event-stream",
]
`;

exports[`shared path between imports 1`] = `
Array [
  "foo",
  "bar",
]
`;

exports[`unclosed block comment 1`] = `Array []`;
